README.md
# tracker部署

```shell
docker run -d --name fastdfs-tracker --net=host -e "FASTDFS_MODE=tracker" -e "FASTDFS_BASE_PATH=/fastdfs/tracker" -v /data/fastdfs/tracker:/fastdfs/tracker -p 22122:22122 fastdfs
```

变量：

| 参数              | 值说明               |
| ----------------- | -------------------- |
| FASTDFS_MODE      | tracker，启动tracker |
| FASTDFS_BASE_PATH | tracker存储路径      |

# storage部署

```shell
docker run -d --name fastdfs-storage --net=host -e "FASTDFS_MODE=storage" -e "FASTDFS_BASE_PATH=/fastdfs/storage" -e "FASTDFS_TRACKER_SERVER_NUM=2" -e "FASTDFS_GROUP_NAME=group1" -e "FASTDFS_TRACKER_SERVER_1=127.0.0.1" -e "FASTDFS_TRACKER_SERVER_2=127.0.0.1"  -e "FASTDFS_STORAGE_PORT=23000" -e "FASTDFS_STORAGE_HTTP_PORT=8888" -v /data/fastdfs/group1/storage:/fastdfs/storage -p 8888:8888 -p 23000:23000 fastdfs
```

变量：

| 参数                       | 值说明                                |
| -------------------------- | ------------------------------------- |
| FASTDFS_MODE               | storage，启动storage和nginx           |
| FASTDFS_TRACKER_SERVER_NUM | tracker节点个数                       |
| FASTDFS_TRACKER_SERVER_1   | tracker节点名字，后面的数据与个数对应 |
| FASTDFS_GROUP_NAME         | fastdfs组名                           |
| FASTDFS_BASE_PATH          | storage存储路径                       |
| FASTDFS_STORAGE_PORT       | storage的server端口号                 |
| FASTDFS_STORAGE_HTTP_PORT  | storage与nginx对应服务的端口号        |

# 访问方式

正常图片：

```http
http://127.0.0.1:8888/group1/M00/00/00/rBIAA1w4WlmEHJ5zAAAAABEZkYo781.png
```

缩略图片：

```http
http://127.0.0.1:8888/group1/M00/00/00/rBIAA1w4WlmEHJ5zAAAAABEZkYo781_200x200.png
```

文件：

```http
http://127.0.0.1:8888/group1/M00/00/00/rBIAA1w4ci-EPmxOAAAAAH644DY0437.gz
```




fastdfs/HISTORY

Version 5.12  2018-06-07
 * code refine for rare case
 * replace print format OFF_PRINTF_FORMAT to PRId64
 * php_ext fix zend_object_store_get_object call in php5.5
 * make.sh uses macros define in /usr/include/fastcommon/_os_define.h
 * correct CRC32, you must upgrade libfastcommon to V1.38 or later

Version 5.11  2017-05-26
 * bug fixed: file_offset has no effect when use trunk file
 * add storage access log header
 * http.conf add parameter http.multi_range.enabed

Version 5.10  2017-03-29
 * use fc_safe_read instead of read, and fc_safe_write instead of write
   you must upgrade libfastcommon to V1.35 or later
 * fix getFileContentEx read bytes,
   you must upgrade libfastcommon to V1.36 or later
 * do NOT sync storage server info to tracker leader
 * adjust parameter store_server when use_trunk_file is true
 * clear sync src id when tracker response ENOENT
 * log more info when fdfs_recv_header / fdfs_recv_response fail

Version 5.09  2016-12-29
 * bug fixed: list_all_groups expand buffer auto for so many groups
 * tracker.conf add parameters: min_buff_size and max_buff_size
 * php extension fix free object bug in PHP 7

Version 5.08  2016-04-08
 * install library to $(TARGET_PREFIX)/lib anyway
 * php extension compiled in PHP 7
 * dio thread use blocked_queue and php extension use php7_ext_wrapper.h,
   you must upgrade libfastcommon to V1.25 or later
 * remove common/linux_stack_trace.[hc]

Version 5.07  2015-09-13
 * schedule task add the "second" field
 * make.sh changed, you must upgrade libfastcommon to V1.21 or later
 * bug fixed: storage_disk_recovery.c skip the first file (binlog first line)
 * bug fixed: should close connection after fetch binlog
 * fdfs_storaged.c: advance the position of daemon_init
 * set log rotate time format
 * bug fixed: must check store_path_index

Version 5.06  2015-05-12
 * compile passed in mac OS Darwin
 * correct scripts in subdir init.d
 * check item thread_stack_size in storage.conf, you must
   upgrade libfastcommon to V1.14 or later

Version 5.05  2014-11-22
 * tracker_mem.c log more info
 * remove useless global variable: g_network_tv
 * storage can fetch it's group_name from tracker server

Version 5.04  2014-09-16
 * add fastdfs.spec for build RPM on Linux
 * depend on libfastcommon
 * in multi tracker servers case, when receive higher status like
   online / active and the storage status is wait_sync or syncing,
   the tracker adjust storage status to newer, and the storage rejoin
   to the tracker server
 * fdfs_monitor support delete empty group
 * bug fixed: two tracker leaders occur in rare case
 * add connection stats
 * delete old log files, add parameter: log_file_keep_days

Version 5.03  2014-08-10
 * network send and recv retry when error EINTR happen
 * support mac OS Darwin
 * use newest logger from libfastcommon
 * patches by liangry@ucweb.com
 * bug fixed: can't sync large files cause by v5.02
 * use newest files from libfastcommon
 * change TRACKER_SYNC_STATUS_FILE_INTERVAL from 3600 to 300
 * socket send and recv ignore erno EINTR

Version 5.02  2014-07-20
 * corect README spell mistake
 * bug fixed: can't deal sync truncate file exception
 * remove tracker_global.c extern keyword to tracker_global.h
 * change log level from ERROR to DEBUG when IOEVENT_ERROR
 * php callback should use INIT_ZVAL to init zval variable
 * add function short2buff and buff2short
 * add get_url_content_ex to support buffer passed by caller
 * logger can set rotate time format
 * logger can log header line
 * #include <stdbool.h> to use C99 bool
 * logger can delete old rotated files
 * bug fixed: connection pool should NOT increase counter when connect fail
 * logger.c do NOT call fsync after write

Version 5.01  2014-02-02
 * trunk binlog be compressed when trunk init
 * bug fixed: sync trunk binlog file to other storage servers immediately when
   the trunk server init done
 * move ioevent_loop.[hc] and fast_task_queue.[hc] from tracker/ to common/
 * hash table support locks
 * hash talbe support new functions: hash_inc and hash_inc_ex

Version 5.00  2013-12-23
 * discard libevent, use epoll in Linux, kqueue in FreeBSD, port in SunOS directly
 * do_notify_leader_changed force close connection when target is myself
 * modify the INSTALL file and tracker/Makefile.in

Version 4.08  2013-11-30
 * bug fixed: FDFS_DOWNLOAD_SERVER_ROUND_ROBIN change to FDFS_STORE_SERVER_ROUND_ROBIN
 * dio_init use memset to init buffer
 * disable linger setting (setsockopt with option SO_LINGER)
 * change log level from error to warning when file not exist on storage server

Version 4.07  2013-06-02
 * make.sh add -lpthread by ldconfig check
 * support multi accept threads
 * tracker and storage server close client connection when recv invalid package
 * client/storage_client.c: file_exist with silence flag
 * tracker and storage process support start, stop and restart command
 * tracker/tracker_proto.c fdfs_recv_header: logDebug change to logError

Version 4.06  2013-01-24
 * fdfs_upload_file tool enhancement
 * fdfs_download_file tool support offset and download size
 * trunk file upload support sub paths rotating correctly
 * add function: fdfs_http_get_file_extension
 * sync truncate file operation anyway

Version 4.05  2012-12-30
 * client/fdfs_upload_file.c can specify storage ip port and store path index
 * add connection pool
 * client load storage ids config
 * common/ini_file_reader.c does NOT call chdir
 * keep the mtime of file same
 * use g_current_time instead of call time function
 * remove embed HTTP support

Version 4.04  2012-12-02
 * bug fixed: get storage server id when storage daemon init
 * storage id in filename use global variable
 * dynamic alloc memory 8 bytes alignment
 * fast_task_queue support memory pool chain

Version 4.03  2012-11-18

libfastcommon/src/md5.c
#include "md5.h"
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static void MD5Transform (UINT4[4], unsigned char[64]);
static void Encode (unsigned char *, UINT4 *, unsigned int);
static void Decode (UINT4 *, unsigned char *, unsigned int);
static void MD5_memcpy (POINTER, POINTER, unsigned int);
static void MD5_memset (POINTER, int, unsigned int);

static unsigned char PADDING[64] = {
	0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/*
 * F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/*
 * ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/*
 * FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4. Rotation is
 * separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
(a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); \
(a) = ROTATE_LEFT ((a), (s)); \
(a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
(a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); \
(a) = ROTATE_LEFT ((a), (s)); \
(a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
(a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); \
(a) = ROTATE_LEFT ((a), (s)); \
(a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
(a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); \
(a) = ROTATE_LEFT ((a), (s)); \
(a) += (b); \
  }

/*
 * MD5 initialization. Begins an MD5 operation, writing a new context.
 */
void 
my_md5_init(MD5_CTX *context)
{
	context->count[0] = context->count[1] = 0;
	/*
	 * Load magic initialization constants.
	 */
	context->state[0] = 0x67452301;
	context->state[1] = 0xefcdab89;
	context->state[2] = 0x98badcfe;
	context->state[3] = 0x10325476;
}

/*
 * MD5 block update operation. Continues an MD5 message-digest operation,
 * processing another message block, and updating the context.
 */
void 
my_md5_update(MD5_CTX        *context, unsigned char  *input, unsigned int    inputLen)
{
	unsigned int    i, index, partLen;

	/* Compute number of bytes mod 64 */
	index = (unsigned int) ((context->count[0] >> 3) & 0x3F);

	/* Update number of bits */
	if ((context->count[0] += ((UINT4) inputLen << 3)) < ((UINT4) inputLen << 3))
		context->count[1]++;
	context->count[1] += ((UINT4) inputLen >> 29);

	partLen = 64 - index;

	/*
	 * Transform as many times as possible.
	 */
	if (inputLen >= partLen) {
		MD5_memcpy
			((POINTER) & context->buffer[index], (POINTER) input, partLen);
		MD5Transform(context->state, context->buffer);

		for (i = partLen; i + 63 < inputLen; i += 64)
			MD5Transform(context->state, &input[i]);

		index = 0;
	} else
		i = 0;

	/* Buffer remaining input */
	MD5_memcpy
		((POINTER) & context->buffer[index], (POINTER) & input[i],
		 inputLen - i);
}

/*
 * MD5 finalization. Ends an MD5 message-digest operation, writing the the
 * message digest and zeroizing the context.
 */
void 
my_md5_final(unsigned char   digest[16], MD5_CTX        *context)
{
	unsigned char   bits[8];
	unsigned int    index, padLen;

	/* Save number of bits */
	Encode(bits, context->count, 8);

	/*
	 * Pad out to 56 mod 64.
	 */
	index = (unsigned int) ((context->count[0] >> 3) & 0x3f);
	padLen = (index < 56) ? (56 - index) : (120 - index);
	my_md5_update(context, PADDING, padLen);

	/* Append length (before padding) */
	my_md5_update(context, bits, 8);
	/* Store state in digest */
	Encode(digest, context->state, 16);

	/*
	 * Zeroize sensitive information.
	 */
	MD5_memset((POINTER) context, 0, sizeof(*context));
}

/*
 * MD5 basic transformation. Transforms state based on block.
 */
static void 
MD5Transform(UINT4           state[4], unsigned char   block[64])
{
	UINT4           a = state[0], b = state[1], c = state[2], d = state[3],
	                x[16];

	Decode(x, block, 64);

	/* Round 1 */
	FF(a, b, c, d, x[0], S11, 0xd76aa478);	/* 1 */
	FF(d, a, b, c, x[1], S12, 0xe8c7b756);	/* 2 */
	FF(c, d, a, b, x[2], S13, 0x242070db);	/* 3 */
	FF(b, c, d, a, x[3], S14, 0xc1bdceee);	/* 4 */
	FF(a, b, c, d, x[4], S11, 0xf57c0faf);	/* 5 */
	FF(d, a, b, c, x[5], S12, 0x4787c62a);	/* 6 */
	FF(c, d, a, b, x[6], S13, 0xa8304613);	/* 7 */
	FF(b, c, d, a, x[7], S14, 0xfd469501);	/* 8 */
	FF(a, b, c, d, x[8], S11, 0x698098d8);	/* 9 */
	FF(d, a, b, c, x[9], S12, 0x8b44f7af);	/* 10 */
	FF(c, d, a, b, x[10], S13, 0xffff5bb1);	/* 11 */
	FF(b, c, d, a, x[11], S14, 0x895cd7be);	/* 12 */
	FF(a, b, c, d, x[12], S11, 0x6b901122);	/* 13 */
	FF(d, a, b, c, x[13], S12, 0xfd987193);	/* 14 */
	FF(c, d, a, b, x[14], S13, 0xa679438e);	/* 15 */
	FF(b, c, d, a, x[15], S14, 0x49b40821);	/* 16 */

	/* Round 2 */
	GG(a, b, c, d, x[1], S21, 0xf61e2562);	/* 17 */
	GG(d, a, b, c, x[6], S22, 0xc040b340);	/* 18 */
	GG(c, d, a, b, x[11], S23, 0x265e5a51);	/* 19 */
	GG(b, c, d, a, x[0], S24, 0xe9b6c7aa);	/* 20 */
	GG(a, b, c, d, x[5], S21, 0xd62f105d);	/* 21 */
	GG(d, a, b, c, x[10], S22, 0x2441453);	/* 22 */
	GG(c, d, a, b, x[15], S23, 0xd8a1e681);	/* 23 */
	GG(b, c, d, a, x[4], S24, 0xe7d3fbc8);	/* 24 */
	GG(a, b, c, d, x[9], S21, 0x21e1cde6);	/* 25 */
	GG(d, a, b, c, x[14], S22, 0xc33707d6);	/* 26 */
	GG(c, d, a, b, x[3], S23, 0xf4d50d87);	/* 27 */
	GG(b, c, d, a, x[8], S24, 0x455a14ed);	/* 28 */
	GG(a, b, c, d, x[13], S21, 0xa9e3e905);	/* 29 */
	GG(d, a, b, c, x[2], S22, 0xfcefa3f8);	/* 30 */
	GG(c, d, a, b, x[7], S23, 0x676f02d9);	/* 31 */
	GG(b, c, d, a, x[12], S24, 0x8d2a4c8a);	/* 32 */

	/* Round 3 */
	HH(a, b, c, d, x[5], S31, 0xfffa3942);	/* 33 */
	HH(d, a, b, c, x[8], S32, 0x8771f681);	/* 34 */
	HH(c, d, a, b, x[11], S33, 0x6d9d6122);	/* 35 */
	HH(b, c, d, a, x[14], S34, 0xfde5380c);	/* 36 */
	HH(a, b, c, d, x[1], S31, 0xa4beea44);	/* 37 */
	HH(d, a, b, c, x[4], S32, 0x4bdecfa9);	/* 38 */
	HH(c, d, a, b, x[7], S33, 0xf6bb4b60);	/* 39 */
	HH(b, c, d, a, x[10], S34, 0xbebfbc70);	/* 40 */
	HH(a, b, c, d, x[13], S31, 0x289b7ec6);	/* 41 */
	HH(d, a, b, c, x[0], S32, 0xeaa127fa);	/* 42 */
	HH(c, d, a, b, x[3], S33, 0xd4ef3085);	/* 43 */
	HH(b, c, d, a, x[6], S34, 0x4881d05);	/* 44 */
	HH(a, b, c, d, x[9], S31, 0xd9d4d039);	/* 45 */
	HH(d, a, b, c, x[12], S32, 0xe6db99e5);	/* 46 */
	HH(c, d, a, b, x[15], S33, 0x1fa27cf8);	/* 47 */
	HH(b, c, d, a, x[2], S34, 0xc4ac5665);	/* 48 */

	/* Round 4 */
	II(a, b, c, d, x[0], S41, 0xf4292244);	/* 49 */
	II(d, a, b, c, x[7], S42, 0x432aff97);	/* 50 */
	II(c, d, a, b, x[14], S43, 0xab9423a7);	/* 51 */
	II(b, c, d, a, x[5], S44, 0xfc93a039);	/* 52 */
	II(a, b, c, d, x[12], S41, 0x655b59c3);	/* 53 */
	II(d, a, b, c, x[3], S42, 0x8f0ccc92);	/* 54 */
	II(c, d, a, b, x[10], S43, 0xffeff47d);	/* 55 */
	II(b, c, d, a, x[1], S44, 0x85845dd1);	/* 56 */
	II(a, b, c, d, x[8], S41, 0x6fa87e4f);	/* 57 */
	II(d, a, b, c, x[15], S42, 0xfe2ce6e0);	/* 58 */
	II(c, d, a, b, x[6], S43, 0xa3014314);	/* 59 */
	II(b, c, d, a, x[13], S44, 0x4e0811a1);	/* 60 */
	II(a, b, c, d, x[4], S41, 0xf7537e82);	/* 61 */
	II(d, a, b, c, x[11], S42, 0xbd3af235);	/* 62 */
	II(c, d, a, b, x[2], S43, 0x2ad7d2bb);	/* 63 */
	II(b, c, d, a, x[9], S44, 0xeb86d391);	/* 64 */

	state[0] += a;
	state[1] += b;
	state[2] += c;
	state[3] += d;

	/*
	 * Zeroize sensitive information.
	 * 
	 */
	MD5_memset((POINTER) x, 0, sizeof(x));
}

/*
 * Encodes input (UINT4) into output (unsigned char). Assumes len is a
 * multiple of 4.
 */
static void 
Encode(unsigned char  *output, UINT4  *input, unsigned int    len)
{
	unsigned int    i, j;

	for (i = 0, j = 0; j < len; i++, j += 4) {
		output[j] = (unsigned char) (input[i] & 0xff);
		output[j + 1] = (unsigned char) ((input[i] >> 8) & 0xff);
		output[j + 2] = (unsigned char) ((input[i] >> 16) & 0xff);
		output[j + 3] = (unsigned char) ((input[i] >> 24) & 0xff);
	}
}

/*
 * Decodes input (unsigned char) into output (UINT4). Assumes len is a
 * multiple of 4.
 */
static void 
Decode(UINT4          *output, unsigned char  *input, unsigned int    len)
{
	unsigned int    i, j;

	for (i = 0, j = 0; j < len; i++, j += 4)
		output[i] = ((UINT4) input[j]) | (((UINT4) input[j + 1]) << 8) |
			(((UINT4) input[j + 2]) << 16) | (((UINT4) input[j + 3]) << 24);
}

/*
 * Note: Replace "for loop" with standard memcpy if possible.
 */

static void 
MD5_memcpy(POINTER         output, POINTER         input, unsigned int    len)
{
	unsigned int    i;

	for (i = 0; i < len; i++)
		output[i] = input[i];
}

/*
 * Note: Replace "for loop" with standard memset if possible.
 */
static void 
MD5_memset(POINTER         output, int             value, unsigned int    len)
{
	unsigned int    i;

	for (i = 0; i < len; i++)
		((char *) output)[i] = (char) value;
}

/*
 * Digests a string
 */
int my_md5_string(char *string,unsigned char digest[16])
{
	MD5_CTX	context;
	unsigned int len = strlen(string);

	my_md5_init(&context);
	my_md5_update(&context, (unsigned char *)string, len);
	my_md5_final(digest, &context);
	return 0;
}

int my_md5_buffer(char *buffer, unsigned int len, unsigned char digest[16])
{
	MD5_CTX	context;

	my_md5_init(&context);
	my_md5_update(&context, (unsigned char *)buffer, len);
	my_md5_final(digest, &context);
	return 0;
}

int my_md5_file(char *filename,unsigned char digest[16])	
{
	FILE           *file;
	MD5_CTX         context;
	int             len;
	unsigned char   buffer[1024];

	if ((file = fopen(filename, "rb")) == NULL)
		return -1;
	else {
		my_md5_init(&context);
		while ((len = fread(buffer, 1, 1024, file)) > 0)
		{
			my_md5_update(&context, buffer, len);
		}
		my_md5_final(digest, &context);

		fclose(file);
	}
	return 0;
}

